### PR TITLE
added ability to pass down url query parameters to initial oauth2.0 authorization page

### PR DESCRIPTION
**Pull request by cosmomeese**
added ability pass down url query parameters to the initial oauth2.0 authorization page. e.g. this is useful for the Fitbit API which defines additional URI parameters (such as token expiration duration, ability to force authorization page/consent and others - see Authorization Page @ https://dev.fitbit.com/build/reference/web-api/oauth2/) that must be appended to the initial authorization URL.

### CODE REVIEW

**Comment by cosmomeese** [comment]
bump? 
What do I do next? This is my first time attempting to contribute to an open source project on github...

**Comment by jennybc** [comment]
Development comes in waves. The "Contribute to active projects" section of [this blog post](http://www.jimhester.com/2017/08/08/contributing/) gives a good sense.

**Comment by cosmomeese** [comment]
Ooooh. This is very helpful - thank you Jenny! :)

**Comment by jennybc** [comment]
httr is on the list but no one's pushing on it right now.

**Comment by cosmomeese** [comment]
Fair enough - like you said, it comes in waves. At least for now I know the tide is out. :P

**Comment by cosmomeese** [comment]
Done - used 'styler' and tweaked manually when it didn't work so well. Let me know if it needs further tweaking.

File: tests/testthat/test-oauth-query-params.R [hunk 1]
```diff
@@ -0,0 +1,67 @@
+context("OAuth-query-params")
+
+test_that("oauth2.0 authorize url appends query params", {
+  app <- oauth_app("x", "y", "z")
+  scope <- httr:::check_scope(c("foo", "bar"))
+  queryParams <- list(
+    foo = "bar",
+    paul = "isCool"
```
Comment (hadley) [inline tests/testthat/test-oauth-query-params.R]: I think you need only a single param here

File: tests/testthat/test-oauth-query-params.R [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  )
+  authURL <- oauth2.0_authorize_url(
+    endpoint = oauth_endpoints("google"),
+    app = app,
+    scope = scope,
+    auth_page_query_params = queryParams
+  )
+
+  url <- parse_url(authURL)
+  paramNames <- names(queryParams)
```
Comment (hadley) [inline tests/testthat/test-oauth-query-params.R]: It would be clearer to write:

```
expect_equal(url$query$foo, "bar")
```

File: tests/testthat/test-oauth-query-params.R [hunk 3]
```diff
@@ -0,0 +1,67 @@
+context("OAuth-query-params")
```
Comment (hadley) [inline tests/testthat/test-oauth-query-params.R]: Please put these tests with the existing OAuth tests

File: tests/testthat/test-oauth-query-params.R [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  url <- parse_url(authURL)
+  paramNames <- names(queryParams)
+  for (qpIndex in 1:length(queryParams))
+  {
+    paramName <- paramNames[qpIndex]
+    expect_equal(url$query[paramName], queryParams[qpIndex])
+  }
+})
+
+test_that("oauth2.0 authorize url does not append empty query params", {
```
Comment (hadley) [inline tests/testthat/test-oauth-query-params.R]: What are you trying to test here? I can't tell because there's so much indirection in your code. Can you please simplify?

File: R/oauth-init.R [hunk 5]
```diff
@@ -142,7 +158,8 @@ oauth2.0_access_token <- function(endpoint,
                                   use_basic_auth = FALSE,
                                   redirect_uri = app$redirect_uri,
                                   client_credentials = FALSE,
-                                  config = list()
+                                  config = list(),
+                                  auth_page_query_params=list()
```
[the remark this answers is not in the document]
Comment (cosmomeese) [reply R/oauth-init.R]: Oversight on my part.

File: tests/testthat/test-oauth-query-params.R (hunk 4 above)
Comment (cosmomeese) [reply tests/testthat/test-oauth-query-params.R]: Trying to test that with an empty/null query_extra parameter the url still looks like it did before my change (i.e. common elements untouched). 
Tried to simplify/clarify - it's nowhere near perfect I just can't think of a more effective way to test for this without testing for each parameter in the authorization URL.

File: tests/testthat/test-oauth.R [hunk 6]
```diff
@@ -53,6 +53,75 @@ test_that("oauth_encode1 works", {
...
+    app = app,
+    scope = scope,
+    query_extra = query_extra
+  )
+  
+  url <- parse_url(authURL)
+  expect_equal(url$query$foo, "bar")
+})
+
+test_that("oauth2.0 authorize url with empty input does not append empty query params", {
```
Comment (hadley) [inline tests/testthat/test-oauth.R]: I still don't understand what you're trying to test here.

Comment (cosmomeese) [reply tests/testthat/test-oauth.R]: basically that providing empty input to `query_extra` doesn't break things.
I've replaced the test with something that's hopefully clearer.

**Comment by cosmomeese** [comment]
Done. Replaced the empty input test with something that's hopefully clearer, more concise and more useful.

**Comment by cosmomeese** [comment]
Done - let me know if it's too wordy (or should be split up into more bullets).

**Comment by hadley** [comment]
Thanks!

**Comment by cosmomeese** [comment]
My pleasure - thanks for the all the help & guidance with this. :)

**Comment by jennybc** [comment]
Hey @cosmomeese, you might like to know this feature comes in handy for Google OAuth flow too. We can pass the user's email address as a hint to the auth server, which pre-selects the targeted identity in the consent screen:

https://github.com/r-lib/gargle/commit/3790e7797bbf82d827b387dcedf0ae79dccc42f9

**Comment by cosmomeese** [comment]
That's so great! Thanks for letting me know - it's always nice to see one's work end up being useful to someone else :)
